scripts/run_all.py

```diff
@@ -4,86 +4,239 @@
+            "examples/python/nyud",
+            "examples/python/text_logging",
+        ]
+
+
+class Viewer:
+    port: int
+    should_close: bool
+    web: bool
+    web_viewer_port: int
+    ws_server_port: int
+    process: Optional[Any]
+
+    def __init__(self, close: bool = False, web: bool = False):
+        self.port = get_free_port()
+        self.should_close = close
+        self.web = web
+        self.web_viewer_port = get_free_port()
+        self.ws_server_port = get_free_port()
+        self.process = None
+
+    def close(self) -> None:
+        if self.process is not None:
+            self.process.kill()
+        self.process = None
+
+    def start(self) -> "Viewer":
+        args = ["--port", str(self.port)]
+        if self.web:
+            args += [
+                "--web-viewer",
+                "--web-viewer-port",
+                str(self.web_viewer_port),
+                "--ws-server-port",
+                str(self.ws_server_port),
+            ]
+
+        self.process = subprocess.Popen(
+            ["cargo", "run", "-p", "rerun", "--all-features", "--"] + args,
+            stdout=subprocess.PIPE,
+        )
+        time.sleep(1)  # give it a moment to start
+        return self
+
+    def __enter__(self) -> "Viewer":
+        self.start()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        if self.process is not None and self.should_close:
+            self.process.kill()
+
+
+def run_sdk_build() -> None:
+    returncode = subprocess.Popen(
+        ["maturin", "develop", "--manifest-path", "rerun_py/Cargo.toml", '--extras="tests"'],
+    ).wait()
+    assert returncode == 0, f"process exited with error code {returncode}"
 
 
-def run_build() -> None:
-    process = subprocess.Popen(
-        ["maturin", "develop", "--manifest-path", "rerun_py/Cargo.toml", '--extras="tests"'],
-    )
-    returncode = process.wait()
+def run_viewer_build() -> None:
+    returncode = subprocess.Popen(["cargo", "build", "-p", "rerun", "--all-features"]).wait()
     assert returncode == 0, f"process exited with error code {returncode}"
 
 
+def run_web(examples: List[str], separate: bool) -> None:
+    if not separate:
+        with Viewer(close=True, web=True) as viewer:
+            for path in examples:
+                run_py_example(path, viewer_port=viewer.port)
+        return
+
+    cleanup: List[Tuple[Any, Any]] = []
+    # start all examples in parallel
+    for path in examples:
+        # each example gets its own viewer
+        viewer = Viewer(web=True).start()
+        example = run_py_example(path, viewer_port=viewer.port, wait=False)
+        cleanup.append((viewer, example))
+
+    # wait for all processes to finish, and close the viewers
+    for pair in cleanup:
+        viewer, example = pair
+        example.wait()
+        viewer.close()
```

Comment:
I think here is we we need to wait a bit more, at least when using a web-viewer.

When we have the web-viewer, the `viewer/Viewer` here is a bit of a misnomer, since it actually refers to the `rerun` process which acts like a triple-server in this case:

* and SDK server, listening to incoming log calls from the log SDK (i.e. from the examples) over TCP (or possibly just loading an pre-made .rrd file)
* a log server, serving the recorded log stream over WebSocket
* a web server, serving the web-viewer wasm and html

in any case; just because the example has finished executing doesn't mean the web-viewer have had enough time to fetch the entire log stream yet.


I think we need something like:

```suggestion
    # Wait for examples to finish logging:
    for example in examples:
        example.wait()
    
    # Give web-viewers (if any) time to fetch all the logged data
    time.sleep(5)
    
    # Shut down the servers/viewers:
    for viewer in viewers:
        viewer.close()
```

Reply:
The `time.sleep(5)` seems to do the trick. I also changed `Viewer.start` to check for log output from the process which states that the server has started listening on the given port. It blocks until it finds that line.

This means that the viewers/servers now start a lot faster, but I don't know if that's necessarily better. The log output is also a lot messier.